### PR TITLE
[Issue #973] handle product is out of stock

### PR DESCRIPTION
**Pull request by ejnshtein**
If a product added to the cart becomes out of stock product will be marked as "Product is out of stock", grayscale filter applied to the image, and if the user tries to click "Secure checkout" it will show error notification.

Screenshot 2020-09-28 at 10 03 06
Screenshot 2020-09-28 at 10 05 19

Same on the Cart page.
Screenshot 2020-09-28 at 10 05 37
Screenshot 2020-09-28 at 10 06 25

### CODE REVIEW

**Comment by yeegor** [comment]
Nice job formatting the pull requests, @ejnshtein. Keep it up!
I am going to take a look now.

**Comment by yeegor** [comment]
Unable to test the changes.
Please provide a step-by-step instruction how did you validate the new logic.

**Comment by ejnshtein** [comment]
How to test:

1. Add existing product to cart.
2. Go to Magento admin panel, find this product and set Stock Status to "Out of Stock"
Screenshot 2020-09-28 at 15 29 53
3. Go back to store view and refresh page.
4. Open CartOverlay
5. See changes

**Review by yeegor — CHANGES_REQUESTED** [review]
Having this in cart:
image

I set the stock status of that product in admin to the following:
image

Save, reload the page and get the following:
image

**Comment by ejnshtein** [comment]
For some reason, the server returns configurable options for out of stock product with `attribute_value: null` that's why the error was thrown.
Ideally, it will be returned as the product is still in the cart, so the user can see which configuration is not available anymore.
Instead, cartData returns the correct product SKU but the variant will not be present in product variants.
Screenshot 2020-09-28 at 16 56 46
Screenshot 2020-09-28 at 16 57 09

Can be treated as another bug, I think.

File: src/app/component/CartItem/CartItem.component.js [hunk 1]
```diff
@@ -287,6 +293,7 @@ export class CartItem extends PureComponent {
                 >
                     <span>{ __('Delete') }</span>
                 </button>
+                { isProductInStock ? (
```
Comment (yeegor) [inline src/app/component/CartItem/CartItem.component.js:296]: Prefer using `if` statements and `renderThing` pattern.

File: src/app/route/CartPage/CartPage.container.js [hunk 2]
```diff
@@ -118,12 +119,34 @@ export class CartPageContainer extends PureComponent {
...
             showNotification,
-            device
+            device,
+            totals
         } = this.props;
 
         // to prevent outside-click handler trigger
         e.nativeEvent.stopImmediatePropagation();
 
+        const hasOutOfStockProductsInCart = totals.items.some((item) => {
```
Comment (yeegor) [inline src/app/route/CartPage/CartPage.container.js:129]: I believe this logic duplicates the one in CartOverlay. Consider separating this logic into an util and keeping it DRY.

**Review by yeegor — CHANGES_REQUESTED** [review]
1. Unable to go to PDP of this product, although the pointer invites me to click.
2. When clicking the card of a disabled product, underlines stick to it and when I move my mouse somewhere else, product name and out of stock notice look like that:
image
3. Also I propose instead of implementing the onclick notification on checkout button, disable it with text alarming the user that he has something out of stock in his cart. I think this is a better way than scare him with red notification.

File: src/app/component/CartItem/CartItem.component.js (hunk 1 above)
Comment (ejnshtein) [reply src/app/component/CartItem/CartItem.component.js:296]: Okay

File: src/app/route/CartPage/CartPage.container.js (hunk 2 above)
Comment (ejnshtein) [reply src/app/route/CartPage/CartPage.container.js:129]: Okay

**Comment by ejnshtein** [comment]
> 1. Unable to go to PDP of this product, although the pointer invites me to click.
> 2. When clicking the card of a disabled product, underlines stick to it and when I move my mouse somewhere else, product name and out of stock notice look like that:
>    image
> 3. Also I propose instead of implementing the onclick notification on checkout button, disable it with text alarming the user that he has something out of stock in his cart. I think this is a better way than scare him with red notification.

1. Should be fixed now.
2. Unable to reproduce.
3. Good idea probably we should do it this way then.

**Comment by yeegor** [comment]
2. Try clicking multiple times and see that the style sticks and the text keeps underlined.
3. Then do it this way please. And let's move it to the next release.

**Review by yeegor — CHANGES_REQUESTED** [review]
Let's stick to the implementation option we agreed on.
We are moving this to the next release.

**Comment by ejnshtein** [comment]
Something like this then?

Screenshot 2020-09-29 at 12 22 30

**Comment by yeegor** [comment]
@ejnshtein it looks amazing in the overlay, but I think we can do better in the cart page.
Maybe it would look better to replace the checkout button with this warning? Or replace its text with the warning's one?
Also, please provide a demonstration on mobile.

**Comment by ejnshtein** [comment]
Final result
Screenshot 2020-09-29 at 14 32 11
Screenshot 2020-09-29 at 14 33 07

**Comment by alfredsgenkins** [comment]
@ejnshtein please make sure not to use conditional rendering so often, this does lead to confusion when reading your code.

**Pull request merged**